### DISCUSSION TITLE
ESModulesLinkingError :(

### DISCUSSION THREAD

**Discussion by binarykitchen** [web-infra-dev/rspack#8309] [Core] [answered]
I'm running a standalone app, using rsbuild, which imports other npm packages compiled by rsbuild as well.

Latest rspack/rsbuild throws this error:
```
ESModulesLinkingError: export 'VideomailClient' (imported as 'VideomailClient') was not found in 'videomail-client' (module has no exports)
```

ESModulesLinkingError?

This npm package [videomail-client](https://github.com/binarykitchen/videomail-client) has this in its package.json:

```json
  "type": "module",
  "exports": {
    "types": "./dist/index.d.ts",
    "import": "./dist/index.js"
  },
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "files": [
    "dist"
  ],
```

See https://www.npmjs.com/package/videomail-client/v/10.0.0-9?activeTab=code

@chenjiahan Since you worked on this error message (git blame tells me), I wonder if you can assist and provide a better explanation to why this can't be compiled?

If you like, I can provide you with access to my repository for reproduction. Thanks.

**Comment by chenjiahan** [comment]
I think the error message is quite clear, the `videomail-client` does not provide a valid export.

For example, try to `import { VideomailClient } from "videomail-client";` in Node.js,  you will get the same error:

![image](https://github.com/user-attachments/assets/7048e9c9-8f3a-4871-ab12-5777986e0904)

**Reply by chenjiahan** [reply, marked as answer]
Note that Rsbuild is not designed to build libraries, you can use Rslib instead.

https://github.com/web-infra-dev/rslib

**Reply by binarykitchen** [reply]
Thanks, I'll check out Rslib tomorrow and let you know ...

Meanwhile ...

> Note that Rsbuild is not designed to build libraries, you can use Rslib instead.

Maybe this warning can be improved to distinguish whether it's an app or a npm package?

**Reply by chenjiahan** [reply]
This warning is thrown by Rspack and normally Rspack follows the ES modules standard to check the modules rather than distinguishing between apps or packages.

**Reply by binarykitchen** [reply]
Ah yeah, the switch to rslib made the difference:
https://github.com/binarykitchen/videomail-client/blob/master/rslib.config.ts

Now that error is gone and I have a new one:
```
[dev:client] start   Building...
[dev:server] ReferenceError: window is not defined
[dev:server]     at /home/michael-heuberger/code/binarykitchen/videomail.io/node_modules/videomail-client/dist/index.js:4369:20
```

I've already set target to `web` so I don't expect that error? Hmmm ....
